pkg/middlewares/replacepath/replace_path.go

```diff
@@ -40,8 +41,22 @@ func (r *replacePath) GetTracingInformation() (string, ext.SpanKindEnum) {
 }
 
 func (r *replacePath) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	req.Header.Add(ReplacedPathHeader, req.URL.Path)
-	req.URL.Path = r.path
+	if req.URL.RawPath == "" {
+		req.Header.Add(ReplacedPathHeader, req.URL.Path)
+	} else {
+		req.Header.Add(ReplacedPathHeader, req.URL.RawPath)
+	}
+
+	req.URL.RawPath = r.path
+
+	var err error
+	req.URL.Path, err = url.PathUnescape(req.URL.RawPath)
```

Comment:
This PR modifies `req.URL.Path` and `req.URL.RawPath` without returning them to the original values. Is that something we want?

Reply:
I don't understand your question, could you elaborate?

Reply:
Look at this test case: https://github.com/containous/traefik/pull/6413/files#diff-d3677828257e31aeca5933a90a133acbR54

In that case, the _original_ request `req.URL.Path` is `/path`, but now is `/foo/bar`, which is neither the requested path, nor the replaced path...

I think this is a behaviour change?

Reply:
What is your concern @dtomcej ? Before the PR, this middleware was mutating req.URL.Path but not req.URL.RawPath. After this PR, it's going to possibly mutate both of them. Are you worried that one of the next handlers in the chain was relying on req.URL.RawPath not being mutated?
imho, since we already meddle with the request and modify req.URL.Path, I think we already conceded that all bets are off for the next handlers. So going one step further and mutating req.URL.RawPath as well is not such a big deal.

Reply:
>Before the PR, this middleware was mutating req.URL.Path but not req.URL.RawPath

Before this pr, the middleware was mutating `req.URL.Path` according to the user-supplied regex replacement, so it ended up what the user expected.

yes, I was thinking for the next step.

Reply:
If something was using `req.URL.RawPath` to know the "ancestor path" it was an error.
And I don't think we do that.